File: src/main/java/offer/Offer61.java
package offer;

import java.util.Arrays;

/**
 * @author jingxinwu
 * @date 2021-11-26 12:57 上午
 */
public class Offer61 {

    public static void main(String[] args) {
        System.out.println(new Offer61().isStraight(new int[]{0, 0, 8, 5, 4}));
        System.out.println(new Offer61().isStraight(new int[]{0, 0, 2, 2, 5}));
    }

    public boolean isStraight(int[] nums) {
        if (nums == null || nums.length == 0) {
            return false;
        }
        Arrays.sort(nums);
        int count = 0;
        for (int num : nums) {

            if (num == 0) {
                count++;
            }
        }
        int diff = 0;
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] == nums[i - 1] && nums[i] != 0) {
                return false;
            }
            if (nums[i] - nums[i - 1] > 1 && nums[i - 1] != 0) {
                diff = diff + nums[i] - nums[i - 1] - 1;
            }
        }
        return diff <= count;
    }
}

File: go/src/leetcode/offer/Offer61.go
package offer

import "sort"

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/3 23:54
 */

func isStraight(nums []int) bool {
	zeroCnt := 0
	sort.Ints(nums)
	for i := 0; i < len(nums)-1; i++ {
		if nums[i] == 0 {
			zeroCnt++
		} else if nums[i] == nums[i+1] {
			return false
		}
	}
	return nums[4]-nums[zeroCnt] < 5
}
